File: python/src/herminas_kernel/contracts.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Protocol, runtime_checkable


class ServiceName(str, Enum):
    CLICKHOUSE = "clickhouse"
    REDPANDA = "redpanda"
    DATA_PLANE = "dataplane"
    INTELLIGENCE = "intelligence"
    API = "api"


class HealthState(str, Enum):
    UNKNOWN = "unknown"
    STARTING = "starting"
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"


@dataclass(frozen=True)
class HealthStatus:
    service: ServiceName
    state: HealthState
    message: str
    checked_at: datetime


@runtime_checkable
class HealthChecker(Protocol):
    def health(self) -> HealthStatus: ...


@dataclass(frozen=True)
class Event:
    """Minimal event envelope shared by ingestion, stream processing and
    storage. Mirrors kernel/contracts/contracts.go's Event and Rust's
    herminas_kernel::contracts::Event."""

    id: str
    dataset: str
    timestamp: datetime
    payload: bytes
    metadata: dict[str, str] = field(default_factory=dict)

File: kernel/contracts/contracts.go
// Package contracts holds the domain types and interfaces shared across
// HermiNas' layers (L0-L3). Rust and Python keep language-native mirrors of
// the same concepts (see rust/kernel/src/contracts.rs and
// python/src/herminas_kernel/contracts.py) until the protobuf contracts
// land in M0.3 (kernel/proto/*.proto) and become the single source of truth
// for cross-process serialization.
package contracts

import (
	"context"
	"time"
)

// ServiceName identifies a service the L2 supervisor manages or probes.
type ServiceName string

const (
	ServiceClickHouse   ServiceName = "clickhouse"
	ServiceRedpanda     ServiceName = "redpanda"
	ServiceDataPlane    ServiceName = "dataplane"
	ServiceIntelligence ServiceName = "intelligence"
	ServiceAPI          ServiceName = "api"
)

// HealthState is the coarse health of a managed service.
type HealthState string

const (
	HealthUnknown   HealthState = "unknown"
	HealthStarting  HealthState = "starting"
	HealthHealthy   HealthState = "healthy"
	HealthDegraded  HealthState = "degraded"
	HealthUnhealthy HealthState = "unhealthy"
)

// HealthStatus is the cross-layer health report: same shape is returned by
// GetHealth (gRPC, M0.3), exposed at /api/v1/health (REST) and printed by
// `herminas status` (CLI, M8).
type HealthStatus struct {
	Service   ServiceName
	State     HealthState
	Message   string
	CheckedAt time.Time
}

// HealthChecker is implemented by anything the supervisor (L2) can probe:
// the embedded engines (ClickHouse, Redpanda) and the Rust/Python runtimes.
type HealthChecker interface {
	Health(ctx context.Context) (HealthStatus, error)
}

// Event is the minimal envelope shared by ingestion, stream processing and
// storage. Kept manually in sync with the Rust and Python equivalents until
// the protobuf schema (kernel/proto/agent.proto, M0.3) replaces it as the
// wire contract.
type Event struct {
	ID        string
	Dataset   string
	Timestamp time.Time
	Payload   []byte
	Metadata  map[string]string
}
